sklearn/metrics/pairwise.py

```diff
@@ -585,6 +585,189 @@ def cosine_distances(X, Y=None):
     return S
 
 
+def gower_distances(X, Y=None, feature_weight=None, categorical_features=None):
+    """Computes the gower distances between X and Y
+
+    Gower is a similarity measure for categorical, boolean and numerical mixed
+    data.
+
+    Parameters
+    ----------
+    X : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    Y : array-like, or pandas.DataFrame, shape (n_samples, n_features)
+
+    feature_weight :  array-like, shape (n_features)
+        According the Gower formula, feature_weight is an attribute weight.
+
+    categorical_features: array-like, shape (n_features)
+        Indicates with True/False whether a column is a categorical attribute.
+        This is useful when categorical atributes are represented as integer
+        values. Categorical ordinal attributes are treated as numeric, and
+        must be marked as false.
+
+        Alternatively, the categorical_features array can be represented only
+        with the numerical indexes of the categorical attribtes.
+
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, n_samples)
+
+    Reference
+    -------
+    Gower, J.C., 1971, A General Coefficient of Similarity and Some of Its
+    Properties.
+
+    Notes
+    ------
+    The non-numeric features, and numeric feature ranges are determined from
+    X and not Y.
+
+    No support for sparse matrices.
+    """
+
+    if issparse(X) or issparse(Y):
+        raise TypeError("Sparse matrices are not supported for gower distance")
+
+    y_none = Y is None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X)
+
+    array_type = np.object
+
+    # this if is necessary as strangelly the validator is rejecting numeric
```

Comment:
This was necessary because of a problem caused by a incorrect treatment of nan values in numeric arrays, in the validation.check_array function.

The validation.check_array function is unable to execute properly, if you pass a float array with a nan value in the middle. You will get this error: ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

You can reproduce this problem with this snippet of code below:

from sklearn.utils import validation
import numpy as np
X=np.array([[21.0,np.nan]])
array_type = np.float64
# the check_array only works if array_type=np.object
# array_type=np.object
validation.check_array(X,dtype=array_type)

To overcome this problem I try all I can to keep the array with its original dtype (line 644), but if the array is numeric, and there are NaNs, the only solution I found is to put a np.object as dtype.

Unless someone has an Idea, I'll keep it as is, and I won't fix this problem in validation, because it seems to be a new ticket, and there will be some impact on the entire framework.